pkg/accounting/accounting.go

```diff
@@ -216,6 +261,45 @@ func (a *Accounting) getPeerBalance(peer swarm.Address) (*PeerBalance, error) {
 	return peerBalance, nil
 }
 
-func (pb *PeerBalance) freeBalance() int64 {
+// expectedBalance returns the balance we expect to have with a peer if all reserved funds are actually credited
+func (pb *PeerBalance) expectedBalance() int64 {
 	return pb.balance - int64(pb.reserved)
 }
+
+// expectedDebt returns the debt we expect to have with a peer if all reserved funds are actually credited
+func (pb *PeerBalance) expectedDebt() uint64 {
+	expectedBalance := pb.expectedBalance()
+	if expectedBalance >= 0 {
+		return 0
+	}
+	return uint64(-expectedBalance)
+}
+
+// NotifyPayment is called by Settlement when we received payment
+func (a *Accounting) NotifyPayment(peer swarm.Address, amount uint64) error {
+	balance, err := a.getPeerBalance(peer)
+	if err != nil {
+		return err
+	}
+
+	balance.lock.Lock()
+	defer balance.lock.Unlock()
+
+	nextBalance := balance.balance - int64(amount)
+
+	// don't allow a payment to put use more into debt than the tolerance
+	if nextBalance < -int64(a.paymentTolerance) {
+		return fmt.Errorf("refusing to accept payment which would put us too much in debt, new balance would have been %d", nextBalance)
```

Comment:
How is this error handled higher-up in the stack? The desired behavior is a disconnect from this peer--and never reconnect again.

Reply:
do we really need to disconnect here?

Reply:
Not needed to disconnect, indeed. 
But still the question: how is this error handled higher up in the stack? Does the peer get notified somehow that you refused the payment?

Reply:
AFAIK as long as its not a disconnect error it just gets logged inside the p2p code (settlement is returning this in its handler).